### PR TITLE
Nans and a feasible bug inisde the `logarithmic_range()`

### PR DESCRIPTION
**Pull request by codemime**
Passing equal values as `min` and `max` args (zero range) to the `logarithmic_range()` results in division by zero inside and a quiet NaN in the returning value. Once assigned to int, it's converted (at least by g++ using Makefile) to a negative number (-2147483648) without any warning. Though that doesn't impact for now, future computations can be ruined easily without anyone noticing quickly (equal limits could have been sent by `player::add_morale()` regularly, but luckily expired morale points are removed). We don't work with a nan =) so there's definitely a feasible bug. I suggest checking and throwing an exception if the check fails. Also, unnecessary computation were avoided.

### CODE REVIEW

**Comment by kevingranade** [comment]
The only part of the game that throws exceptions is the json module; this
is a choice not an oversight.
In cases like this we sanitize the output (clearly it should return 0) if
possible, and print a debug message if not.

**Comment by codemime** [comment]
Sorry. I've checked exception policy before adding the throw and found some throwing, but didn't spot that it's json-related. Now I see. I suppose that if an invalid range is passed here, then the caller's logic is somehow broken. You just can't normalize anything, given an invalid range. Third arg is useless.
Shouldn't we also add a debugmsg() here to be able to track those cases?

**Comment by kevingranade** [comment]
That's what I meant by print a debug message, yes.

**Comment by codemime** [comment]
All set.

File: src/cata_utility.cpp [hunk 1]
```diff
@@ -193,10 +194,16 @@ double logarithmic_range( int min, int max, int pos )
     const double LOGI_MIN = logarithmic( -LOGI_CUTOFF );
     const double LOGI_MAX = logarithmic( +LOGI_CUTOFF );
     const double LOGI_RANGE = LOGI_MAX - LOGI_MIN;
-    // Anything beyond [min,max] gets clamped.
-    if( pos < min ) {
+
+    if( min >= max ) {
+        debugmsg( "Invalid range" );
```
Comment (Coolthulhu) [inline src/cata_utility.cpp]: That `debugmsg` could have more details, like: "logarithmic_range: minimum >= maximum ([value of minimum] >= [value of maximum])".

Also, requiring `min` to be strictly lesser than `max` (not allowing equality) could break something. Morale code uses it.

Comment (codemime) [inline src/cata_utility.cpp]: > That debugmsg could have more details

You're right. I thought that macro debugmsg() does most of that and it actually does. It writes: file name, line number and no function name (weird) into a log file only. I found, that some files use DebugLog() along with debugmsg() to report an error, others don't. What's the most appropriate way of reporting and logging errors you'd use?

> Also, requiring min to be strictly lesser than max (not allowing equality) could break something. Morale code uses it.

It won't break anything for now, I've taken it into account (see the first comment).

> equal limits could have been sent by player::add_morale() regularly, but luckily expired morale_points are removed.

**Comment by codemime** [comment]
I've done what @Coolthulhu has suggested.

**Pull request merged**